### PR TITLE
Halls endpoints

### PR DESCRIPTION
**Pull request by DarkJediDJ**
# Description
This PR adds halls endpoints

## Checklist
- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No error nor warning in the console.

## Items
- [x] Added 3 new endpoints;
- [x] Added route handler;
- [x] Added DB queries;
- [x] Fixed migrations naming;

### CODE REVIEW

File: api/halls/halls.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	"fmt"
+	"net/http"
+
+	halldb "github.com/darkjedidj/cinema-service/internal/repository/halls"
+	"github.com/gorilla/mux"
+)
+
+var Repo halldb.Repository
+
+func Handle(response http.ResponseWriter, request *http.Request) {
```
Comment (gordonBusyman) [inline api/halls/halls.go]: Usually http verbs are used for functionality separation. Use them here

File: api/halls/halls.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	vars := mux.Vars(request)
+	route := vars["route"]
+	switch route {
+	case "add":
+		CreateHall(response, request)
```
Comment (gordonBusyman) [inline api/halls/halls.go]: `halls.CreateHall() ` we did speak about it

File: internal/repository/halls/hall.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	"database/sql"
+)
+
+type Repository struct {
+	DB *sql.DB
+}
+
+type Resource struct {
+	ID  int  `json:"ID"`
+	VIP bool `json:"VIP"`
```
Comment (gordonBusyman) [inline internal/repository/halls/hall.go]: Seats amount?

File: api/halls/halls.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		DeleteHall(response, request)
+	default:
+		response.WriteHeader(http.StatusBadGateway)
+	}
+}
+
+func CreateHall(response http.ResponseWriter, request *http.Request) {
+	response.Header().Set("Content-Type", "application/json")
+	var hall halldb.Resource
+	json.NewDecoder(request.Body).Decode(&hall)
```
Comment (gordonBusyman) [inline api/halls/halls.go]: Check for errors!

File: cmd/cinetickets/main.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	if e != nil {
+		log.Fatal("Error loading .env file")
+	}
+	psqlconn := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", os.Getenv("DB_HOST"), os.Getenv("DB_PORT"), os.Getenv("DB_USER"), os.Getenv("DB_PASSWORD"), os.Getenv("DB_NAME"))
+	db, err := sql.Open("postgres", psqlconn)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	halls.Repo.DB = db
```
Comment (gordonBusyman) [inline cmd/cinetickets/main.go]: Remove it here and initialize as we discussed.

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: cmd/cinetickets/main.go (hunk 5 above)
Comment (DarkJediDJ) [reply cmd/cinetickets/main.go]: can u repeat please, cause i need to open connection in main and transfer it somehow into the other package, and that is the best way i discovered

File: api/halls/halls.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+var Repo halldb.Repository
+
+//Handle handles all endpoints on this route
+func Handle(response http.ResponseWriter, request *http.Request) {
+	vars := mux.Vars(request)
+	route := vars["route"]
+	switch route {
+	case "post":
```
Comment (gordonBusyman) [inline api/halls/halls.go]: WTF is this?

File: api/halls/halls.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	if err != nil {
+		response.WriteHeader(http.StatusBadGateway)
+	}
+}
+
+//Delete get json and deletes Hall with the same ID
+func Delete(response http.ResponseWriter, request *http.Request) {
+	response.Header().Set("Content-Type", "application/json")
+	var hall halldb.Resource
+	json.NewDecoder(request.Body).Decode(&hall)
```
Comment (gordonBusyman) [inline api/halls/halls.go]: Do I need to point each place where you SHOULD check for errors?
What does `.Decode()` return?

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: api/halls/halls.go (hunk 6 above)
Comment (DarkJediDJ) [reply api/halls/halls.go]: couldnt understand what do u mean? naming? switch case usage? way i get parameters?

Comment (gordonBusyman) [reply api/halls/halls.go]: Describe the endpoints which you are going to handle here

File: cmd/cinetickets/main.go (hunk 5 above)
Comment (gordonBusyman) [reply cmd/cinetickets/main.go]: Server will have db property, then router, then handler, then service, then repository

File: api/halls/halls.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+var Repo halldb.Repository
+
+//Handle handles all endpoints on this route
+func Handle(db *sql.DB) http.HandlerFunc {
+	return func(response http.ResponseWriter, request *http.Request) {
+		switch request.Method {
+		case http.MethodPost:
+			Create(response, request, db) //[POST] BASE_URL/v1/halls/create
```
Comment (gordonBusyman) [inline api/halls/halls.go]: Comments

File: api/halls/halls.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+	err = Repo.Delete(int64(id))
+	if err != nil {
+		response.WriteHeader(http.StatusBadGateway)
+	}
+
+	response.WriteHeader(http.StatusOK)
+}
+
+//Get get json and selects Hall with the same ID
```
Comment (gordonBusyman) [inline api/halls/halls.go]:
```suggestion
// Get get json and selects Hall with the same ID
```

File: api/server.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	myRouter.HandleFunc("/v1/halls/{id}", halls.Handle(db))
+	a.Router = myRouter
+}
+
+// Run starts server
+func (a *App) Run(addr string) {
+	log.Fatal(http.ListenAndServe(addr, a.Router))
+}
+
+func (a *App) ConnectDB() *sql.DB{
```
Comment (gordonBusyman) [inline api/server.go]: Move it to main.. as we said before

File: internal/repository/halls/hall.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
+package halldb
```
Comment (gordonBusyman) [inline internal/repository/halls/hall.go]:
```suggestion
package hall
```

File: internal/repository/halls/hall.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//Retrieve Hall from DB
+func (r *Repository) Retrieve(id int64) (dbHall Resource, e error) {
+
+	rows, err := r.DB.Query(`SELECT * FROM halls WHERE "id" = $1`, id)
```
Comment (gordonBusyman) [inline internal/repository/halls/hall.go:30]: Try squirell

File: internal/repository/halls/hall.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+func (r *Repository) Create(hall Resource) error {
+	insertHall := `insert into halls("vip","seats") values($1,$2)`
+	_, err := r.DB.Exec(insertHall, hall.VIP, hall.Seats)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//Retrieve Hall from DB
```
Comment (gordonBusyman) [inline internal/repository/halls/hall.go]:
```suggestion
// Retrieve Hall from DB
```
https://go.dev/doc/effective_go

File: api/halls/halls.go [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+	"database/sql"
+	"encoding/json"
+	"log"
+	"net/http"
+	"strconv"
+
+	halldb "github.com/darkjedidj/cinema-service/internal/repository/halls"
+	"github.com/gorilla/mux"
+)
+
```
Comment (gordonBusyman) [inline api/halls/halls.go:15]:
```
type Handler struct {
    Repo: halldb.Repository
}
```

File: api/server.go [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+type App struct {
+	Router *mux.Router
+}
+
+// New creates router with handler
+func (a *App) New(db *sql.DB) {
+
+	myRouter := mux.NewRouter().StrictSlash(false)
+	myRouter.HandleFunc("/v1/halls/{id}", halls.Handle(db))
```
Comment (gordonBusyman) [inline api/server.go]:
```suggestion
	myRouter.HandleFunc("/v1/halls/{id}", &halls.Handler{&hall.Repo{db}}.Handle())
```

I know that this is not working, but move to this direction

File: api/halls/halls.go (hunk 14 above)
Comment (DarkJediDJ) [reply api/halls/halls.go:15]: what does ":" mean? didnt see such syntax, and why not using pointer on Repository struct there?

File: api/server.go (hunk 15 above)
Comment (DarkJediDJ) [reply api/server.go]: like, i pushed my vision of solving it, but the error doesnt match problem in the code i guess 
image

Comment (DarkJediDJ) [reply api/server.go]: yeah, and i cant understand why we are using "&" before first package, why we are taking adress, we appeal to the method, not to the pointer

File: internal/repository/halls/hall.go (hunk 12 above)
Comment (DarkJediDJ) [reply internal/repository/halls/hall.go:30]: i need to test DB forwarding with working connection, then will try use sqirell, cause last time it didnt work. And i guess debug errors in db connection and server logic at the same time isnt very easy, wanna try solve one problem, and then try other

Comment (gordonBusyman) [reply internal/repository/halls/hall.go:30]: Ok, leave it for now. I've created a [ticket](https://github.com/DarkJediDJ/cinetickets/projects/1#card-76777466), which is first in your queue

File: api/halls/halls.go (hunk 14 above)
Comment (gordonBusyman) [reply api/halls/halls.go:15]: Use pointer, ignore ":", i was writing it while doing smth else

File: api/server.go (hunk 15 above)
Comment (gordonBusyman) [reply api/server.go]: The solution is not about syntax, what I was trying to say is having a struct for the handler, where u have a struct for the Repo, which contains the db

File: internal/repository/halls/hall.go [hunk 16]
```diff
@@ -14,6 +14,8 @@ type Resource struct {
 	Seats int  `json:"seats"`
 }
 
+var Repo Repository
```
Comment (gordonBusyman) [inline internal/repository/halls/hall.go]: Do u use it? What is this here?

**Review by gordonBusyman — CHANGES_REQUESTED** [review]

File: internal/repository/halls/hall.go (hunk 16 above)
Comment (DarkJediDJ) [reply internal/repository/halls/hall.go]: just tried different options, nvm

File: api/server.go [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+type App struct {
+	Router *mux.Router
+}
+
+// New creates router with handler
+func (a *App) New(db *sql.DB) {
+
+	myRouter := mux.NewRouter().StrictSlash(false)
+	handler := halls.Handler{Repo: &hall.Repository{DB: db}}
+	myRouter.HandleFunc("/v1/halls/{id}", handler.HandleID)
```
Comment (gordonBusyman) [inline api/server.go]: I know that gorilla/mux has something called Subrouters, i am not sure that it can help here, but check plz, it might help

Comment (DarkJediDJ) [reply api/server.go]: Routes can be used as subrouters: nested routes are only tested if the
  parent route matches. This is useful to define groups of routes that
  share common conditions like a host, a path prefix or other repeated
  attributes. As a bonus, this optimizes request matching.

 I guess its usefull, but wont solve problem with 2 Handle methods

**Review by gordonBusyman — APPROVED** [review]

**Pull request merged**